lib/config.js

```diff
@@ -0,0 +1,204 @@
+// Copyright 2017 TODO Group. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+const Ajv = require('ajv')
+const fetch = require('node-fetch')
+const findConfig = require('find-config')
+const fs = require('fs')
+const jsonfile = require('jsonfile')
+const lodash = require('lodash')
+const path = require('path')
+const yaml = require('js-yaml')
+
+const Rules = require('../rules/rules')
+const RuleInfo = require('./ruleinfo')
+const Fixes = require('../fixes/fixes')
+
+function isAbsoluteURL(url) {
+  try {
+    const u = new URL(url)
+    if (u.host !== '') {
+      return true
+    }
+  } catch (e) {}
+  return false
+}
+
+/**
+ * Load a ruleset config from the specified location.
+ *
+ * @param {string} [configLocation] A URL, file location, or directory containing a repolinter config file
+ * @param {array} [processed] list of config files already processed, used to prevent loops
+ * @returns {Object} The loaded repolinter json config
+ * @throws Will throw an error if unable to parse config or if config is invalid
+ */
+async function loadConfig(configLocation, processed = []) {
+  if (configLocation == null) {
+    throw new Error('must specify config location')
+  }
+
+  let configData = null
+  if (isAbsoluteURL(configLocation)) {
+    const res = await fetch(configLocation)
+    if (!res.ok) {
+      throw new Error(
+        `Failed to fetch config from ${configLocation} with status code ${res.status}`
+      )
+    }
+    configData = await res.text()
+  } else {
+    const st = await fs.promises.stat(configLocation)
```

Comment:
Instead of specifying a directory, I think we should specify a root filename and let the `extends` parameter specify where to traverse next, since that way the user can control the layout in the configuration files w/o changing what they specify to the CLI. That being said, we still need to put this logic somewhere: maybe add another parameter `targetDirectory`, and search `targetDirectory` for the config if `configLocation` is `null`?

Reply:
I don't entirely follow what you mean by root filename, and the relationship here with `extends`.  

I did debate whether to pass `targetDirectory` as a separate parameter, but determined I didn't necessarily need to if I used stat like this, but I'm not opposed to keeping those separate.  I was trying to go for the simplest possible API for `loadConfig`, which was basically "just tell me where to look, and I'll try to figure out what you meant".

Reply:
Hmm... what if config file discovery was separated out from the `loadConfig` function?  So the logic inside of `repolinter.lint` becomes something like:

``` javascript
if (!ruleset) {
  // use find-config module to find config files in targetDirectory, or fallback to default.yaml
  ruleset = config.findConfig(targetDirectory)
}
config.loadConfig(ruleset)
```

Then `loadConfig` is always handed an explicit file or URL and doesn't have to deal with any of the directory logic, nor does it need to know about `targetDirectory` at all.

Reply:
Sure, that sounds good. It'll definitely be a lot cleaner than what's there now😅.

Reply:
separated `findConfig` from `loadConfig`.